File: Algorithms/p019_Remove_Nth_Node_From_End_of_List/p019_Remove_Nth_Node_From_End_of_List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode** pp = &head;
        ListNode* p = head;
        for(;n>1;--n){
            p = p->next;
        }
        while(p->next){
            pp = &(*pp)->next;
            p = p->next;
        }
        // save and free *pp
        *pp = (*pp)->next;
        return head;
    }
};

File: Algorithms/p019_Remove_Nth_Node_From_End_of_List/p019_Remove_Nth_Node_From_End_of_List.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def removeNthFromEnd(self, head, n):
        """
        :type head: ListNode
        :type n: int
        :rtype: ListNode
        """
        h = ListNode(None)
        h.next = head
        p1 = h
        p2 = head
        while n>1:
            p2 = p2.next
            n -= 1
        while p2.next:
            p1 = p1.next
            p2 = p2.next
        p1.next = p1.next.next
        return h.next
